R/do_svd.R

```diff
@@ -0,0 +1,308 @@
+      component <- seq(min(length(variance), n_component))
+      if (output=="wide") {
+        # this return one row data frame with d of svd
+        mat <- matrix(variance[component], ncol=length(component))
+        ret <- as.data.frame(mat)
+        colnames(ret) <- avoid_conflict(c(grouped_col), paste("axis", seq(ncol(mat)), sep=""))
+      } else if (output=="long") {
+        # this return a data frame with a column with d of svd
+        ret <- data.frame(component = component, svd.value = variance[component])
+        colnames(ret) <- avoid_conflict(grouped_col, c("new.dimension", value_cname))
+      } else {
+        stop(paste(output, "is not supported as output"))
+      }
+    } else {
+      stop(paste(type, "is not supported as type argument."))
+    }
+    ret
+  }
+
+  # Calculation is executed in each group.
+  # Storing the result in this tmp_col and
+  # unnesting the result.
+  # If the original data frame is grouped by "tmp",
+  # overwriting it should be avoided,
+  # so avoid_conflict is used here.
+  tmp_col <- avoid_conflict(grouped_col, "tmp")
+  ret <- df %>%
+    dplyr::do_(.dots=setNames(list(~do_svd_each(.)), tmp_col)) %>%
+    unnest_with_drop_(tmp_col)
+
+  ret
+}
+
+#' Calculate distance of each pair of groups.
+#' @param df data frame in tidy format
+#' @param group A column you want to calculate the correlations for.
+#' @param dimension A column you want to use as a dimension to calculate the correlations.
+#' @param value A column for the values you want to use to calculate the correlations.
+#' @param distinct The returned pair should be duplicated in swapped order or not.
+#' TRUE makes it easy to filter group names.
+#' @param diag If similarity between itself should be returned or not.
+#' @param method Type of calculation. https://cran.r-project.org/web/packages/proxy/vignettes/overview.pdf
+#' @export
+do_svd.cols <- function(df,
+                         ...,
+                        type="group",
+                        fill=0,
+                        fun.aggregate=mean,
+                        n_component=3,
+                        centering=TRUE,
+                        output ="long"){
+  loadNamespace("dplyr")
+  loadNamespace("tidyr")
+  loadNamespace("reshape2")
+  loadNamespace("stats")
+  loadNamespace("lazyeval")
+
+  grouped_col <- grouped_by(df)
+  label_col <- col_name(substitute(label))
+
+  select_dots <- lazyeval::lazy_dots(...)
+
+  value_colname <- avoid_conflict(grouped_col, "value")
+
+  # this is executed on each group
+  do_svd_each <- function(df){
+    # create matrix by selected columns
+    matrix <- df %>%
+      dplyr::select_(.dots=select_dots) %>%
+      as.matrix() %>%
+      na.omit()
+
+    if(centering){
+      # move the origin to center of data
+      matrix <- sweep(matrix, 2, colMeans(matrix), "-")
+    }
+    if(type=="group"){
+      # u matrix in svd can be regarded as coordinations of groups in reduced dimension
```

Comment:
took some time to realize result$u is u matrix.
comment like, "u matrix (result$u) in svd can..." would help.